### ISSUE TITLE
RUST_BACKTRACE=1 incorrectly reports _main as the name of the symbol before _main in the binary

### ISSUE THREAD

**Issue by roy-work** [rust-lang/rust#68674]
The last / earliest frame in a `RUST_BACKTRACE=1` output is named incorrectly; it appears the correct name should be `_main` or `main`, but the symbol I get is always whatever symbol precedes `_main` in the actual binary.

Note that I'm referring to the low-level entrypoint of the binary, not the `fn main()` that the user writes. That `fn main()` also appears in the trace, in the correct place and named correctly, but it is not the last frame, as several other frames precede it, which is the Rust runtime setting itself up. 

For example, if I run my current project with `RUST_BACKTRACE=1`, it panics.

(the actual cause of the panic is my fault; this bug is only about the output that `RUST_BACKTRACE` emits after that crash)

We get<sup>1</sup>:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NamedSubprocessError { subprocess_name: "ssh", error: Failed(1) }', src/libcore/result.rs:1165:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:139
  14: core::result::unwrap_failed
             at src/libcore/result.rs:1165
  15: core::result::Result<T,E>::unwrap
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/result.rs:933
  16: project_name::main
             at [snip]/src/main.rs:139
  17: std::rt::lang_start::{{closure}}
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  18: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:48
  19: std::panicking::try::do_call
             at src/libstd/panicking.rs:287
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  21: std::panicking::try
             at src/libstd/panicking.rs:265
  22: std::panic::catch_unwind
             at src/libstd/panic.rs:396
  23: std::rt::lang_start_internal
             at src/libstd/rt.rs:47
  24: std::rt::lang_start
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  25: <project_name::_IMPL_DESERIALIZE_FOR_NameOfAStructThatImpsDeserialize::<impl serde::de::Deserialize for project_name::NameOfAStructThatImpsDeserialize>::deserialize::__Visitor as serde::de::Visitor>::expecting
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Note these frames in particular:
```
  16: project_name::main
             at [snip]/src/main.rs:139
  17: std::rt::lang_start::{{closure}}
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  24: std::rt::lang_start
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  25: <project_name::_IMPL_DESERIALIZE_FOR_NameOfAStructThatImpsDeserialize::<impl serde::de::Deserialize for project_name::NameOfAStructThatImpsDeserialize>::deserialize::__Visitor as serde::de::Visitor>::expecting
```
That is, my `fn main()` is being called by `lang_start`; that's expected. `lang_start` is being called by … a serde derive? That's not expected.

If I run this in `lldb`, and put a breakpoint: `(lldb) b core::panicking::panic_fmt`, and then it pauses. `lldb` does not appear confused about where we are, and shows this frame as:
```
(lldb) up
frame #12: 0x000000010000cb02 project-name`std::rt::lang_start::he84252c74d1ae0a2(main=(project-name`project_name::main::hf1a57fdc1eddf2ac at main.rs:53), argc=13, argv=0x00007ffeefbff078) at rt.rs:61:4
(lldb) up
frame #13: 0x000000010000a832 project-name`main + 34
```

So, LLDB seems to think that frame is `main` (and that's a C name, not a Rust name). Reasonable if it is the entrypoint to the binary.

And if we do some disassesmbly:
```
(lldb) disassemble --frame
project-name`main:
    0x10000a810 <+0>:  pushq  %rbp
    0x10000a811 <+1>:  movq   %rsp, %rbp
    0x10000a814 <+4>:  subq   $0x10, %rsp
    0x10000a818 <+8>:  movslq %edi, %rax
    0x10000a81b <+11>: leaq   -0x6112(%rip), %rdi       ; project_name::main::hf1a57fdc1eddf2ac at main.rs:53
    0x10000a822 <+18>: movq   %rsi, -0x8(%rbp)
    0x10000a826 <+22>: movq   %rax, %rsi
    0x10000a829 <+25>: movq   -0x8(%rbp), %rdx
    0x10000a82d <+29>: callq  0x10000cac0               ; std::rt::lang_start::he84252c74d1ae0a2 at rt.rs:58
->  0x10000a832 <+34>: addq   $0x10, %rsp
    0x10000a836 <+38>: popq   %rbp
    0x10000a837 <+39>: retq
```
This is sort of what I expect for an entrypoint: we load the address of my `fn main`, and call into `lang_start`; this is definitely a binary entrypoint.

Here, we're suspecting that `RUST_BACKTRACE=1` is just choosing some other symbol name for the frame. If we disassemble more of the binary:

```
__ZN207_$LT$project_name.._IMPL_DESERIALIZE_FOR_NameOfAStructThatImpsDeserialize..$LT$impl$u20$serde..de..Deserialize$u20$for$u20$project_name..NameOfAStructThatImpsDeserialize$GT$..deserialize..__Visitor$u20$as$u20$serde..de..Visitor$GT$9expecting17he4cce3311b14248cE:
000000010000a7d0        pushq   %rbp
000000010000a7d1        movq    %rsp, %rbp
000000010000a7d4        subq    $0x20, %rsp
000000010000a7d8        leaq    0x47555f(%rip), %rax
000000010000a7df        movq    %rdi, -0x10(%rbp)
000000010000a7e3        movq    %rsi, -0x8(%rbp)
000000010000a7e7        movq    -0x8(%rbp), %rdi
000000010000a7eb        movq    %rax, %rsi
000000010000a7ee        movl    $0x19, %edx
000000010000a7f3        callq   __ZN57_$LT$core..fmt..Formatter$u20$as$u20$core..fmt..Write$GT$9write_str17h68360bed214ad6ccE ## _$LT$core..fmt..Formatter$u20$as$u20$core..fmt..Write$GT$::write_str::h68360bed214ad6cc
000000010000a7f8        movb    %al, -0x11(%rbp)
000000010000a7fb        movb    -0x11(%rbp), %al
000000010000a7fe        andb    $0x1, %al
000000010000a800        movzbl  %al, %eax
000000010000a803        addq    $0x20, %rsp
000000010000a807        popq    %rbp
000000010000a808        retq
000000010000a809        nopl    (%rax)
_main:
000000010000a810        pushq   %rbp
000000010000a811        movq    %rsp, %rbp
000000010000a814        subq    $0x10, %rsp
000000010000a818        movslq  %edi, %rax
000000010000a81b        leaq    __ZN14project_name4main17hf1a57fdc1eddf2acE(%rip), %rdi ## project_name::main::hf1a57fdc1eddf2ac
000000010000a822        movq    %rsi, -0x8(%rbp)
000000010000a826        movq    %rax, %rsi
000000010000a829        movq    -0x8(%rbp), %rdx
000000010000a82d        callq   __ZN3std2rt10lang_start17he84252c74d1ae0a2E ## std::rt::lang_start::he84252c74d1ae0a2
000000010000a832        addq    $0x10, %rsp
000000010000a836        popq    %rbp
000000010000a837        retq
000000010000a838        nop
000000010000a839        nop
000000010000a83a        nop
000000010000a83b        nop
000000010000a83c        nop
```

Here, we can see that the weird symbol that `RUST_BACKTRACE=1` outputs does sort of correlate with the entrypoint: it's the symbol immediately before `_main` (`main` in LLDB). So, if `RUST_BACKTRACE=1` is ignoring non-Rust symbols when it figures out what to print… then that would explain why that frame is wrong.

A minimal example suffices as well:
```rust
use std::io;

fn main() {
    Err(io::Error::new(io::ErrorKind::InvalidData, "error")).unwrap()
}
```

And a minimal `Cargo.toml`:
```toml
[package]
name = "the_crate_name"
version = "0.0.1"
edition = "2018"
```

```
» RUST_BACKTRACE=1 cargo run
   Compiling the_crate_name v0.0.1 (/Users/roy/code/random/bt2)
    Finished dev [unoptimized + debuginfo] target(s) in 1.19s
     Running `target/debug/the_crate_name`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "error" }', src/libcore/result.rs:1165:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:188
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:205
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:464
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:373
  12: rust_begin_unwind
             at src/libstd/panicking.rs:302
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:139
  14: core::result::unwrap_failed
             at src/libcore/result.rs:1165
  15: core::result::Result<T,E>::unwrap
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libcore/result.rs:933
  16: the_crate_name::main
             at src/main.rs:4
  17: std::rt::lang_start::{{closure}}
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  18: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:48
  19: std::panicking::try::do_call
             at src/libstd/panicking.rs:287
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  21: std::panicking::try
             at src/libstd/panicking.rs:265
  22: std::panic::catch_unwind
             at src/libstd/panic.rs:396
  23: std::rt::lang_start_internal
             at src/libstd/rt.rs:47
  24: std::rt::lang_start
             at /rustc/73528e339aae0f17a15ffa49a8ac608f50c6cf14/src/libstd/rt.rs:61
  25: the_crate_name::main
```

Again, `the_crate_name::main` in the last frame (frame 21) should perhaps read `main` or `_main`: it's not `the_crate_name::main` that's running there; `the_crate_name::main` is really frame 12. In a minimal example, we have very few symbols to choose from.

Again, the disassembly shows that `_main` is preceded by the symbol being output, here, `the_crate_name::main`:

<details>
  <summary>Click to expand disassembly</summary>
<pre><code>__ZN14the_crate_name4main17h445f977752289da6E:
0000000100001cb0        pushq   %rbp
0000000100001cb1        movq    %rsp, %rbp
0000000100001cb4        subq    $0x30, %rsp
0000000100001cb8        leaq    0x1f55c(%rip), %rax
0000000100001cbf        movb    $0xc, -0x1(%rbp)
0000000100001cc3        leaq    -0x18(%rbp), %rdi
0000000100001cc7        movzbl  -0x1(%rbp), %esi
0000000100001ccb        movq    %rax, %rdx
0000000100001cce        movl    $0x5, %ecx
0000000100001cd3        callq   __ZN3std2io5error5Error3new17h7ad5cd9b8a04b0efE ## std::io::error::Error::new:
:h7ad5cd9b8a04b0ef
0000000100001cd8        movq    -0x18(%rbp), %rax
0000000100001cdc        movq    %rax, -0x28(%rbp)
0000000100001ce0        movq    -0x10(%rbp), %rax
0000000100001ce4        movq    %rax, -0x20(%rbp)
0000000100001ce8        leaq    -0x28(%rbp), %rdi
0000000100001cec        callq   __ZN4core6result19Result$LT$T$C$E$GT$6unwrap17h9094cf16ad18ae1eE ## core::resu
lt::Result$LT$T$C$E$GT$::unwrap::h9094cf16ad18ae1e
0000000100001cf1        addq    $0x30, %rsp
0000000100001cf5        popq    %rbp
0000000100001cf6        retq
0000000100001cf7        nopw    (%rax,%rax)
_main:
0000000100001d00        pushq   %rbp
0000000100001d01        movq    %rsp, %rbp
0000000100001d04        subq    $0x10, %rsp
0000000100001d08        movslq  %edi, %rax
0000000100001d0b        leaq    __ZN14the_crate_name4main17h445f977752289da6E(%rip), %rdi ## the_crate_name::main::h445f977752289da6
0000000100001d12        movq    %rsi, -0x8(%rbp)
0000000100001d16        movq    %rax, %rsi
0000000100001d19        movq    -0x8(%rbp), %rdx
0000000100001d1d        callq   __ZN3std2rt10lang_start17hc62f9e31cf02e7fdE ## std::rt::lang_start::hc62f9e31cf02e7fd
0000000100001d22        addq    $0x10, %rsp
0000000100001d26        popq    %rbp
0000000100001d27        retq
0000000100001d28        nop
0000000100001d29        nop
0000000100001d2a        nop
0000000100001d2b        nop
</code></pre>
</details>

**OS:** OS X Mojave
**`rustc --version`:** `rustc 1.40.0 (73528e339 2019-12-16)`

<sub><sup>1</sup>I've changed symbol names to remove the actual name of the project, as it relates to my employment. In the mangled symbol names, I have not updated the character counts, so they're not correct mangle-ings anymore.</sub>

**Comment by alexcrichton** [comment]
A bit late to report on this, but if someone runs into this can the `backtrace` crate be tested out? A few bugs have been fixed as-of-late and I'm curious if it reports the right version of the backtrace here.

**Comment by Enselic** [comment]
Triage: Let's close since there has been no reports about this still being a problem.
